examples/cst/cst_NodeSelector.py

```diff
@@ -73,66 +91,78 @@ def visit_Assign(self, node: cst.Assign):
-            func_name = node.func.value
-            if self.is_imported_from_pandas(func_name):
-                return self.create_ir_node_for(func_name, args, kwargs)
+        if isinstance(node.func, cst.Name):  # calls to directly imported library function. e.g. read_sql
+            func_alias = node.func.value
+            if func_alias not in self.library_methods:
+                return
+            module, func_name = self.library_methods[func_alias]
+            result = self.call_module_method(module, func_name, args, kwargs)
 
-        if isinstance(node.func, cst.Attribute):
+        elif isinstance(node.func, cst.Attribute):
             attribute = self.generic_visit(node.func.value)
-            if isinstance(attribute, cst.Name):
-                if attribute.value in self.pandas_aliases:
-                    func_name = node.func.attr.value
-                    return self.create_ir_node_for(func_name, args, kwargs)
-            if isinstance(attribute, DataFrameNode):
-                func_name = node.func.attr.value
-                return self.create_ir_node_for(func_name, [attribute] + args, kwargs)
 
-    def visit_Name(self, node: cst.Name):
+            if isinstance(attribute, cst.Name):  # calls to function accessed via module. e.g. pd.read_sql
+                if attribute.value not in self.libraries:
+                    return
+                module = self.libraries[attribute.value]
+                func_name = node.func.attr.value
+                result = self.call_module_method(module, func_name, args, kwargs)
+            elif isinstance(attribute, IRNode):  # calls to IR nodes. e.g. df.join
+                func_name = node.func.attr.value
+                node = attribute
+                module = module_by_name[node.library]
+                result = self.call_module_ir_method(module, node, func_name, args, kwargs)
+            else:
+                print("Warning: Something strange got called: ", attribute)
+
+        if result:
+            return result
+
+    def call_module_method(self, module: InputModule, func_name: str, args: list, kwargs: dict) -> Node:
+        result = module.visit_call(func_name, args, kwargs)
+        if result:
+            if not result.library:
+                result.library = module.module_name
+            return result
+        raise NotImplementedError(f"Rewrite rule for '{func_name}' of '{module.module_name}' is not implemented.")
+
+    def call_module_ir_method(
+        self, module: InputModule, ir_node: IRNode, func_name: str, args: list, kwargs: dict
+    ) -> Node:
+        result = module.visit_call_on_ir_node(ir_node, func_name, args, kwargs)
+        if result:
+            if not result.library:
+                result.library = module.module_name
+            return result
+        raise NotImplementedError(
+            f"Rewrite rule for '{func_name}' on IR object for '{module.module_name}' is not implemented."
+        )
```

Comment:
Hm also die Ähnlichkeit zwischen call_module_method und call_module_ir_method gefällt mit wirklich nicht, aber es es ist nicht trivial das in einer methode zu machen, da funktionen wie df1.join(df2) und pd.join(df1, df2) (falls es so eine Methode in einer Bibliothek geben würde) mit dem altem Ansatz verwechelt werden würden

Reply:
Ja die Methode gibts: pd.merge()

Finde ich erstmal OK so. Den Namen finde ich ein bisschen nichtssagend, aber yolo